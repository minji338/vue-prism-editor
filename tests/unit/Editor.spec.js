import { mount } from "@vue/test-utils";
import Editor from "@/components/Editor.vue";
import "prismjs";
// jest.mock("prismjs");
import { compileToFunctions } from "vue-template-compiler";

global.window.getSelection = jest.fn(() => ({}));
describe("Editor.vue", () => {
  it("renders", () => {
    const code = "initialCode";
    const wrapper = mount(Editor, {
      propsData: { code }
    });
    expect(wrapper.html()).toContain(code);
  });
  it("sets contentEditable", () => {
    const wrapper = mount(Editor);
    const $pre = wrapper.find("pre");
    expect($pre.attributes().contenteditable).toBe("true");
    wrapper.setProps({
      readonly: true
    });
    expect($pre.attributes().contenteditable).toBe("false");
  });
  it("emits change event", () => {
    const code = "console.log('test')";
    const wrapper = mount(Editor);
    const $pre = wrapper.find("pre");
    $pre.element.innerHTML = code;
    $pre.trigger("keyup");
    expect(wrapper.emitted("change")[0]).toEqual([code]);
  });

  it("v-model works", () => {
    const compiled = compileToFunctions(
      '<div><Editor class="foo" v-model="code" /></div>'
    );
    const wrapper = mount(compiled, {
      data: () => ({
        code: "test"
      }),
      stubs: {
        Editor
      }
    });
    const $pre = wrapper.find("pre");

    expect(wrapper.vm.code).toEqual("test");
    $pre.element.innerHTML = "works";
    wrapper.vm.$children[0].$emit("change", "works");
    expect(wrapper.vm.code).toEqual("works");
  });

  it("emits keydown event", () => {
    const mockHandler = jest.fn();
    const code = "console.log('test')";
    const wrapper = mount(Editor, {
      propsData: {
        emitEvents: true
      },
      listeners: {
        keydown: mockHandler
      }
    });
    const $pre = wrapper.find("pre");
    $pre.element.innerHTML = code;
    $pre.trigger("keydown");
    expect(wrapper.emitted()["keydown"]).toBeTruthy();
    expect(mockHandler).toHaveBeenCalled();
  });

  it("renders with null value", () => {
    const code = null;
    const wrapper = mount(Editor, {
      propsData: { code }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("emits keyup event", () => {
    const mockHandler = jest.fn();
    const code = "console.log('test')";
    const wrapper = mount(Editor, {
      propsData: {
        emitEvents: true
      },
      listeners: {
        keyup: mockHandler
      }
    });
    const $pre = wrapper.find("pre");
    $pre.element.innerHTML = code;
    $pre.trigger("keyup");
    expect(wrapper.emitted()["keyup"]).toBeTruthy();
    expect(mockHandler).toHaveBeenCalled();
  });

  it("not neccessary but", async () => {
    const code = "log()";
    const wrapper = mount(Editor, {
      propsData: {
        code
      },
      sync: false
    });

    expect(wrapper.vm.content).toMatchInlineSnapshot(
      `<code class="language-js"><span class="token function">log</span><span class="token punctuation">(</span><span class="token punctuation">)</span></code>`
    );
  });
});
