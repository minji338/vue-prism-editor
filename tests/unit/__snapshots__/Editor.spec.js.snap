// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor.vue renders with null value 1`] = `
<div class="prism-editor-wrapper">
  <!----> <pre contenteditable="true" spellcheck="false" autocapitalize="off" autocomplete="off" autocorrect="off" data-gramm="false" class="prism-editor__code language-js"><code class="language-js"></code></pre>
</div>
`;
